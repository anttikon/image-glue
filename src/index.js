import sharp from 'sharp'

async function getMetadata(file) {
  const { width, height, channels, format, hasAlpha } = await sharp(file).metadata()
  return { width, height, channels, format, hasAlpha }
}

async function createFile(file1, file2, opts) {
  const { output } = opts
  let image = sharp({ create: opts })[opts.format](output)
  image = sharp(await image.toBuffer()).overlayWith(file1, { gravity: sharp.gravity.west })[opts.format](output)
  image = sharp(await image.toBuffer()).overlayWith(file2, { gravity: sharp.gravity.east })[opts.format](output)
  return image[opts.format](output).toBuffer()
}

function getOptions(metadata1, metadata2, opts) {
  return {
    width: metadata1.width + metadata2.width,
    height: metadata1.height,
    channels: metadata1.channels,
    format: opts.format ? opts.format : metadata1.format,
    output: opts.output || {},
    background: opts.background ? opts.background : metadata1.hasAlpha ? { r: 0, g: 0, b: 0, alpha: 0 } : {
      r: 255,
      g: 255,
      b: 255
    }
  }
}

async function resizeImage(file, height) {
  const resizedFile = await sharp(file).resize(null, height).toBuffer()
  const metadata = await getMetadata(resizedFile)
  return { file: resizedFile, metadata }
}

export const merge = async (files, opts = {}) => {
  if (files.filter(f => !!f).length !== 2) {
    throw new Error('merge should be called with two parameters')
  }

  const [metadata1, metadata2] = await Promise.all([getMetadata(files[0]), getMetadata(files[1])])

  if (metadata1.height > metadata2.height) {
    const { file, metadata } = await resizeImage(files[0], metadata2.height)
    return createFile(file, files[1], getOptions(metadata, metadata2, opts))
  } else if (metadata2.height > metadata1.height) {
    const { file, metadata } = await resizeImage(files[1], metadata1.height)
    return createFile(files[0], file, getOptions(metadata1, metadata, opts))
  } else {
    return createFile(files[0], files[1], getOptions(metadata1, metadata2, opts))
  }
}
