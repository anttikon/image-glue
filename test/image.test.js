import fs from 'fs'
import { merge } from '../src/index'
import path from 'path'
import sharp from 'sharp'

async function expectImage(file) {
  const { width, height, format, hasAlpha, channels } = await sharp(file).metadata()
  expect({ width, height, format, hasAlpha, channels }).toMatchSnapshot()
}

describe('merge', () => {

  beforeAll(() => {
    if (!fs.existsSync(path.resolve(__dirname, 'results'))) {
      fs.mkdirSync(path.resolve(__dirname, 'results'))
    }
  })

  it('merge equal size images', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_sun.jpg'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_v.jpg'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2])
    fs.writeFileSync(path.resolve(__dirname, 'results/50x50_sun_v.jpg'), mergedFile)
    await expectImage(mergedFile)
  })

  it('merge smaller image with larger image', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_sun.jpg'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_beer.jpg'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2])
    fs.writeFileSync(path.resolve(__dirname, 'results/50x50_sun_beer.jpg'), mergedFile)
    await expectImage(mergedFile)
  })

  it('merge larger image with smaller image', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_donut.jpg'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_sun.jpg'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2])
    fs.writeFileSync(path.resolve(__dirname, 'results/50x50_donut_sun.jpg'), mergedFile)
    await expectImage(mergedFile)
  })

  it('from transparent', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_donut.png'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_sun.jpg'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2])
    fs.writeFileSync(path.resolve(__dirname, 'results/50x50_donut_sun.png'), mergedFile)
    await expectImage(mergedFile)
  })

  it('to transparent', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_v.jpg'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_beer.png'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2])
    fs.writeFileSync(path.resolve(__dirname, 'results/50x50_v_beer.jpg'), mergedFile)
    await expectImage(mergedFile)
  })

  it('should require two parameters', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_v.jpg'))
    await expect(merge([file1])).rejects.toEqual(new Error('merge should be called with two parameters'))
    await expect(merge([null])).rejects.toEqual(new Error('merge should be called with two parameters'))
    await expect(merge([file1, null])).rejects.toEqual(new Error('merge should be called with two parameters'))
    await expect(merge([null, file1])).rejects.toEqual(new Error('merge should be called with two parameters'))
  })

  it('target filetype', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/50x50_v.jpg'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_beer.png'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2], { format: 'png' })
    fs.writeFileSync(path.resolve(__dirname, 'results/50x50_v_beer.png'), mergedFile)
    await expectImage(mergedFile)
  })
  it('target background color', async () => {
    const file1 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_donut.png'))
    const file2 = fs.readFileSync(path.resolve(__dirname, 'images/100x100_beer.png'))

    await expectImage(file1)
    await expectImage(file2)

    const mergedFile = await merge([file1, file2], { format: 'png', background: { r: 150, g: 150, b: 150 } })
    fs.writeFileSync(path.resolve(__dirname, 'results/100x100_donut_beer.png'), mergedFile)
    await expectImage(mergedFile)
  })

})