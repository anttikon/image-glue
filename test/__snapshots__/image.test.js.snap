// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge from transparent 1`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 100,
  "width": 100,
}
`;

exports[`merge from transparent 2`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge from transparent 3`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 50,
  "width": 100,
}
`;

exports[`merge merge equal size images 1`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge merge equal size images 2`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge merge equal size images 3`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 100,
}
`;

exports[`merge merge equal size images with transparency 1`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 50,
  "width": 50,
}
`;

exports[`merge merge equal size images with transparency 2`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 50,
  "width": 50,
}
`;

exports[`merge merge equal size images with transparency 3`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 50,
  "width": 100,
}
`;

exports[`merge merge larger image with smaller image 1`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 100,
  "width": 100,
}
`;

exports[`merge merge larger image with smaller image 2`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge merge larger image with smaller image 3`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 100,
}
`;

exports[`merge merge smaller image with larger image 1`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge merge smaller image with larger image 2`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 100,
  "width": 100,
}
`;

exports[`merge merge smaller image with larger image 3`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 100,
}
`;

exports[`merge target background color 1`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 100,
  "width": 100,
}
`;

exports[`merge target background color 2`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 100,
  "width": 100,
}
`;

exports[`merge target background color 3`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 100,
  "width": 200,
}
`;

exports[`merge target filetype 1`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge target filetype 2`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 100,
  "width": 100,
}
`;

exports[`merge target filetype 3`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 50,
  "width": 100,
}
`;

exports[`merge to transparent 1`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 50,
}
`;

exports[`merge to transparent 2`] = `
Object {
  "channels": 4,
  "format": "png",
  "hasAlpha": true,
  "height": 100,
  "width": 100,
}
`;

exports[`merge to transparent 3`] = `
Object {
  "channels": 3,
  "format": "jpeg",
  "hasAlpha": false,
  "height": 50,
  "width": 100,
}
`;
